pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -120,34 +127,47 @@ rec {
         ]
       ];
 
+      # When cross-compiling checks need to be enabled explicitly via crossAttrs
+      doCheck' = if stdenv.hostPlatform == stdenv.buildPlatform
+                 then doCheck
+                 else attrs.crossAttrs.doCheck or false;
```

Comment:
So still needs to be `doCheck = doCheck && (stdenv.hostPlatform == stdenv.buildPlatform);` for now, but `git grep 'doCheck = true; # not cross'` for what can be removed once we do `config.doCheckByDefault or true` (!).

Reply:
Honestly, I think we should just move `crossAttrs` merging from whatever file it is now to `mkDerivation`. Having an standard way to override for cross builds in the derivation itself is nice.

And the question of `doCheck = true;` by default in cross builds is a separate issue (as you note in at the top of the review).

I imagine the migration path to #33599 do be like this:

- This PR (everything's optional, disabled on cross).
- Many fixes.
- `doInstallCheck = true` by default on non-cross.
- Some more fixes.
- `doCheck = true` by default on non-cross.
- Some more fixes.
- Then the cross questions.

Reply:
`crossAttrs` is bad because we want to cross packages to have as little cross-specific behavior as possible. I *want* the awakward `++ stdenv.lib.optionals (hostPlatform != buildPlatform) ...` hijinkses there as a red-flag that things aren't as they should be. If you see how it is used in the wild, you will see that it indeed mask all sorts of nasty duplication and the like.

Now yes the `doCheck = hostPlatform = buildPlatform;` thing is very common, and less of a red flag, so my hope with the issue that once `config.doCheckByDefault` becomes on by default, those can just be emitted altogether. It's a trick to encode *just* this cross special-case by default to avoid needing to mess with `doCheck`. Until we are ready to turn on `config.doCheckByDefault`, I rather keep the existing behavior.

[Real real long term, install checks should be their own derivation so we can run regardless of cross.]

Reply:
Currently

- no `doCheck` in a derivation means "do the default of config.doCheckByDefault, don't check on cross",
- `doCheck = false` means "broken",
- `doCheck = true` means "not broken on non-cross" (a)

What you propose is

- no `doCheck` in a derivation means "doCheck non-cross",
- `doCheck = false` means "broken",
- `doCheck = true` means "works both on cross and not" (b)

The problem is that

- currently `doCheck = true` means (a) and you want it to mean (b), changing semantics this way in this PR will break things, and I want this PR to break nothing,

> Now yes the doCheck = hostPlatform = buildPlatform; thing is very common, and less of a red flag, so my hope with the issue that once config.doCheckByDefault becomes on by default, those can just be emitted altogether

- I'm not sure we want `config.doCheckByDefault` by default, 2x slowdown is still a lot. It would be ok to set it by default if we make another option like `config.doCheckEvenIfExpensiveByDefault` and mark expensive things like atlas and llvm tests. I would not recommend it to be set by default for all the things, though.

See the new patch.